library/fluid/fluid_extensions/src/main/java/org/quiltmc/qsl/fluid/api/FlowableFluidExtensions.java

```diff
@@ -0,0 +1,372 @@
+	// and only made available to cover as many cases as possible.
+
+	default void spawnSplashParticles(Entity splashing, Vec3d splashPos, RandomGenerator random) {
+		for (int i = 0; i < 1.0f + splashing.getDimensions(splashing.getPose()).width * 20.0f; ++i) {
+			double xOffset = (random.nextDouble() * 2.0 - 1.0) * (double) splashing.getDimensions(splashing.getPose()).width;
+			double zOffset = (random.nextDouble() * 2.0 - 1.0) * (double) splashing.getDimensions(splashing.getPose()).width;
+			int yFloor = MathHelper.floor(splashing.getY());
+			ParticleEffect particle = splashParticle(splashing, splashPos, random);
+			if (particle != null) {
+				splashing.world.addParticle(
+						particle,
+						splashing.getX() + xOffset,
+						yFloor + 1.0f,
+						splashing.getZ() + zOffset,
+						splashPos.x,
+						splashPos.y,
+						splashPos.z
+				);
+			}
+		}
+	}
+
+	default void spawnBubbleParticles(Entity splashing, Vec3d splashPos, RandomGenerator random) {
+		for (int i = 0; i < 1.0f + splashing.getDimensions(splashing.getPose()).width * 20.0f; ++i) {
+			double xOffset = (random.nextDouble() * 2.0 - 1.0) * (double) splashing.getDimensions(splashing.getPose()).width;
+			double zOffset = (random.nextDouble() * 2.0 - 1.0) * (double) splashing.getDimensions(splashing.getPose()).width;
+			int yFloor = MathHelper.floor(splashing.getY());
+			ParticleEffect particle = bubbleParticle(splashing, splashPos, random);
+			if (particle != null) {
+				splashing.world.addParticle(
+						particle,
+						splashing.getX() + xOffset,
+						yFloor + 1.0f,
+						splashing.getZ() + zOffset,
+						splashPos.x,
+						splashPos.y - random.nextDouble() * 0.2,
+						splashPos.z
+				);
+			}
+
+		}
+	}
+
+	default void onSplash(World world, Vec3d pos, Entity splashing, RandomGenerator random) {
+		Entity passenger = splashing.hasPassengers() && splashing.getPrimaryPassenger() != null ? splashing.getPrimaryPassenger() : splashing;
+		float volumeMultiplier = passenger == splashing ? 0.2f : 0.9f;
+		Vec3d velocity = passenger.getVelocity();
+		double volume = Math.min(
+				1.0f,
+				Math.sqrt(velocity.x * velocity.x * 0.2 + velocity.y * velocity.y + velocity.z * velocity.z * 0.2D) * volumeMultiplier
+		);
+
+		SoundEvent sound = volume < 0.25f ? splashSound(splashing, pos, random) : highSpeedSplashSound(splashing, pos, random);
+		if (sound != null) {
+			splashing.playSound(
+					sound,
+					(float) volume,
+					1.0f + (random.nextFloat() - random.nextFloat()) * 0.4f
+			);
+		}
+
+		spawnBubbleParticles(splashing, pos, random);
+		spawnSplashParticles(splashing, pos, random);
+
+		GameEvent splash = splashGameEvent(splashing, pos, random);
+		if (splash != null) {
+			splashing.emitGameEvent(splash);
+		}
+	}
+
+	/**
+	 * Here you can modify viscosity and speed based on enchantments.
+	 *
+	 * @return a float array where [0] holds updated horizontal viscosity, and [1] holds updated speed.
```

Comment:
_Please_ make this return a class with named fields. I know we don't have tuples in Java-land, but magic array indexes are worse.